### PR TITLE
feat: authentication - Create authentication apis

### PR DESCRIPTION
**Pull request by RiteshAdwani**
This PR introduces functionalities for user authentication in our social media backend and also some changes in the boiler plate code. The key features include:

**sm-16 - Backend boiler plate initial setup**
1. Implemented centralized environment variable management for improved configuration and type safety.
2. Updated .env.sample and validateRequest middleware
3. Replace status codes with enums

**sm-26 - User Registration**
1. File upload handling with Multer for the profile picture.
2. Integration with Cloudinary for secure and efficient image storage.
3. Data validation using Zod schemas, ensuring that incoming data meets our requirements.
5. Proper database interaction to save the new user.
6. Generation of both access and refresh tokens upon successful registration.
7. Sending back a well-structured response with user data and access token along with refresh token in cookie.

**sm-27 - User Login**
1. Checking the validity of user's email and password.
2. Generation of both access and refresh tokens upon successful login.
3. Sending back a well-structured response with user data and access token along with refresh token in response cookie.

**sm-28 - User Logout**
1. Verification of access token with verifyToken middleware.
2. Clearing refresh token from response cookie. 

**sm-33 - Refresh Token**
1. Generation of new access token after verification of refresh token with verifyToken middleware

**sm-34 - Forgot Password**
1. Generation of reset password token after checking if user exists or not
2. Integration of nodemailer to send mails
3. Sending reset password link to the user

**sm-35 - Reset Password**
1. Verification of the reset password token
2. Hashing of new password and updation in database

### CODE REVIEW

File: src/middlewares/verifyToken.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import jwt from 'jsonwebtoken';
+import { ACCESS_TOKEN_SECRET_KEY, REFRESH_TOKEN_SECRET_KEY } from '../utils/env-variables';
+
+const verifyToken = (tokenType: string) => {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    // Extract the token from the request headers
+    const token = req.headers.authorization?.split(' ')[1];
+
+    if (!token) {
+      return res.status(StatusCodes.UNAUTHORIZED).json({
```
Comment (VatsalDave2102) [inline src/middlewares/verifyToken.ts:12]: Error response could be thrown using errorHandler middleware

File: src/middlewares/verifyToken.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        success: false,
+        message: 'No token provided',
+        data: null,
+      });
```
Comment (VatsalDave2102) [inline src/middlewares/verifyToken.ts]: Error response could be thrown using errorHandler middleware

**Review by VatsalDave2102 — CHANGES_REQUESTED** [review]

File: src/utils/token.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * Generates an access token for the given user
+ * @param {JwtUserPayload} user - User payload for JWT
+ * @returns {Object} Object containing access token and expiration time
+ */
+const generateAccessToken = (user: JwtUserPayload) => {
+  const accessTokenSecretKey = ACCESS_TOKEN_SECRET_KEY;
+  // TODO: Update expiration time to 1 hour
+  const accessTokenExpiryDuration = 60 * 1; // 1 minute (for testing)
+  if (!accessTokenSecretKey) {
+    throw new AppError('ACCESS TOKEN SECRET KEY is not defined', StatusCodes.INTERNAL_SERVER_ERROR);
```
Comment (VatsalDave2102) [inline src/utils/token.ts]: Could add a condition to show token errors only in development environment

**Review by VatsalDave2102 — APPROVED** [review]
Looks good, approved.

**Review by kashyap27102 — APPROVED** [review]

**Review by svipulc — APPROVED** [review]

**Pull request merged**